### PR TITLE
API Access to Game State

### PR DESCRIPTION
**Pull request by jwarwick**
Creation of an api-key to allow 3rd party extensions to access a user's game state.

### CODE REVIEW

File: src/clj/web/lobby.clj [hunk 1]
```diff
@@ -254,13 +254,14 @@
 (defn handle-lobby-create
   [{{{:keys [username] :as user} :user} :ring-req
     client-id                           :client-id
-    {:keys [title format allow-spectator save-replay spectatorhands password room side]} :?data :as event}]
+    {:keys [title format allow-spectator api-access save-replay spectatorhands password room side]} :?data :as event}]
```
Comment (merrickbrown) [inline src/clj/web/lobby.clj:257]: Maybe this happens elsewhere/as a matter of course, but should we check the that the user has-api-keys on the backend against the value passed from the frontend?

Comment (jwarwick) [reply src/clj/web/lobby.clj:257]: I'm really only using the `has-keys` to make the UI clearer (so someone doesn't check the box and not know how to make it work). Any real authorization will happen when a request comes in on the (yet-to-be-written) API.

Comment (jwarwick) [reply src/clj/web/lobby.clj:257]: Maybe another point that isn't clear - you need to have an API key and check the box in the Create Game screen to make the API work. So you can create games without allowing access with the API key.

File: src/clj/web/api.clj [hunk 2]
```diff
@@ -106,8 +113,8 @@
 
 (defroutes routes
   (wrap-routes private-routes wrap-anti-forgery)
-  (wrap-routes public-CSRF-routes wrap-anti-forgery)
-  public-routes)
+  public-routes
```
Comment (NoahTheDuke) [inline src/clj/web/api.clj:116]: Any reason for this?

File: src/cljs/nr/gamelobby.cljs [hunk 3]
```diff
@@ -528,7 +531,20 @@
...
+         (let [has-keys (:has-api-keys @user false)]
+           [:p
+            [:label
+             [:input {:disabled (not has-keys)
+                      :type "checkbox" :checked (:api-access @s)
+                      :on-change #(swap! s assoc :api-access (.. % -target -checked))}]
+             (tr [:lobby.api-access "Allow API access to game information"])
+             (when (not has-keys)
+               (str " " (tr [:lobby.api-requires-key "(Requires an API Key in Settings)"])))
+             ]])
```
Comment (NoahTheDuke) [inline src/cljs/nr/gamelobby.cljs:545]: Hanging parens

**Review by NoahTheDuke — APPROVED** [review]
This is very interesting! Thanks for doing all this work! Can't wait to see bots on jnet

File: src/clj/web/api.clj (hunk 2 above)
Comment (jwarwick) [reply src/clj/web/api.clj:116]: The `public-CSRF-routes` has this route handler `(GET "/*" [] pages/index-page)`, which blocks the `GET` in `public-routes`. You still want it with the CSRF token, so pulling it out doesn't make a ton of sense.

Comment (NoahTheDuke) [reply src/clj/web/api.clj:116]: That makes sense.

**Pull request merged**